src/target/puya.c

```diff
@@ -0,0 +1,234 @@
+	return true;
+}
+
+static bool puya_flash_prepare(target_flash_s *flash)
+{
+	target_mem32_write32(flash->t, PUYA_FLASH_KEYR, PUYA_FLASH_KEYR_KEY1);
+	target_mem32_write32(flash->t, PUYA_FLASH_KEYR, PUYA_FLASH_KEYR_KEY2);
+
+	uint8_t hsi_fs =
+		(target_mem32_read32(flash->t, PUYA_RCC_ICSCR) >> PUYA_RCC_ICSCR_HSI_FS_SHIFT) & PUYA_RCC_ICSCR_HSI_FS_MASK;
+	if (hsi_fs > 4)
+		hsi_fs = 0;
+	DEBUG_TARGET("HSI frequency selection is %d\n", hsi_fs);
+
+	const uint32_t eppara0 = target_mem32_read32(flash->t, PUYA_FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 0);
+	const uint32_t eppara1 = target_mem32_read32(flash->t, PUYA_FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 4);
+	const uint32_t eppara2 = target_mem32_read32(flash->t, PUYA_FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 8);
+	const uint32_t eppara3 = target_mem32_read32(flash->t, PUYA_FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 12);
+	const uint32_t eppara4 = target_mem32_read32(flash->t, PUYA_FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 16);
+	DEBUG_TARGET("PY32 flash timing cal 0: %08" PRIx32 "\n", eppara0);
+	DEBUG_TARGET("PY32 flash timing cal 1: %08" PRIx32 "\n", eppara1);
+	DEBUG_TARGET("PY32 flash timing cal 2: %08" PRIx32 "\n", eppara2);
+	DEBUG_TARGET("PY32 flash timing cal 3: %08" PRIx32 "\n", eppara3);
+	DEBUG_TARGET("PY32 flash timing cal 4: %08" PRIx32 "\n", eppara4);
+
+	target_mem32_write32(flash->t, PUYA_FLASH_TS0, eppara0 & 0xffU);
+	target_mem32_write32(flash->t, PUYA_FLASH_TS1, (eppara0 >> 16U) & 0x1ffU);
+	target_mem32_write32(flash->t, PUYA_FLASH_TS3, (eppara0 >> 8U) & 0xffU);
+	target_mem32_write32(flash->t, PUYA_FLASH_TS2P, eppara1 & 0xffU);
+	target_mem32_write32(flash->t, PUYA_FLASH_TPS3, (eppara1 >> 16U) & 0x7ffU);
+	target_mem32_write32(flash->t, PUYA_FLASH_PERTPE, eppara2 & 0x1ffffU);
+	target_mem32_write32(flash->t, PUYA_FLASH_SMERTPE, eppara3 & 0x1ffffU);
+	target_mem32_write32(flash->t, PUYA_FLASH_PRGTPE, eppara4 & 0xffffU);
+	target_mem32_write32(flash->t, PUYA_FLASH_PRETPE, (eppara4 >> 16U) & 0x3fffU);
+
+	return true;
+}
+
+static bool puya_flash_done(target_flash_s *flash)
+{
+	target_mem32_write32(flash->t, PUYA_FLASH_CR, PUYA_FLASH_CR_LOCK);
+	return true;
+}
+
+static bool puya_wait_flash(target_s *const target, platform_timeout_s *const timeout)
+{
+	while (target_mem32_read32(target, PUYA_FLASH_SR) & PUYA_FLASH_SR_BSY) {
+		if (target_check_error(target))
+			return false;
+		if (timeout)
+			target_print_progress(timeout);
+	}
+	return true;
+}
+
+static bool puya_check_flash_no_error(target_s *const target)
+{
+	uint32_t status = target_mem32_read32(target, PUYA_FLASH_SR);
+	if (status & PUYA_FLASH_SR_WRPERR) {
+		DEBUG_ERROR("puya flash erase error: sr 0x%" PRIx32 "\n", status);
+		return false;
+	}
+	return true;
+}
+
+static bool puya_flash_erase(target_flash_s *flash, target_addr_t addr, size_t len)
+{
+	(void)len;
+	target_mem32_write32(flash->t, PUYA_FLASH_CR, PUYA_FLASH_CR_PER);
+	target_mem32_write32(flash->t, addr, 0);
+	if (!puya_wait_flash(flash->t, NULL)) {
+		return false;
+	}
```

Comment:
Please remove the excess braces here and the check below - they're not necessary and should be dropped.